docs/notebooks/Hidden_Linear_Function_Problem.ipynb

```diff
@@ -0,0 +1,527 @@
+    "\n",
+    "#### 2. Quantum circuits have advantage over classical when solving this problem\n",
+    "\n",
+    "As we will show below, this problem can be solved with a Clifford circuit. Therefore, according to the [Gottesman–Knill theorem](https://en.wikipedia.org/wiki/Gottesman%E2%80%93Knill_theorem), this problem can be solved in polynomial time on a classical computer. So, it might look like quantum computers aren't better than classical ones in solving this problem.\n",
+    "\n",
+    "However, if we apply certain restrictions on matrix $A$, the circuit will have fixed depth (i.e. number of Moments). Namely, if the matrix $A$ is an adjacency matrix of a \"grid\" graph (whose edges can be colored in 4 colors), all CZ gates will fit in 4 moments, and overall we will have only 8 moments - and this doesn't depend on $n$.\n",
+    "\n",
+    "But for classical circuits it can be proven (see [1]) that even if we restrict matrix $A$ in the same way, the depth of classical circuit (with gates of bounded fan-in) must grow as $n$ grows (in fact, it grows as $\\log(n)$).\n",
+    "\n",
+    "### Structure of this notebook\n",
+    "\n",
+    "In the first part, we solve this problem in a straightforward way, which follows directly from the problem definition. We will use this solution to verify our quantum solution in the next part. Also this part contains helper code to generate \"interesting\" instances of the problem.\n",
+    "\n",
+    "In the next part we solve this problem with Cirq. First, we write a code which builds a quantum circuit for solving arbitrary instances of the problem. Then we use Cirq's Clifford simulator to simulate this circuit. We do it for small instances and compare results to the bruteforce solution from the previous part. Then we solve the problem for a larger instance of the problem to demonstrate that it can be solved efficiently.\n",
+    "\n",
+    "Goal of this notebook is to introduce the reader to the problem and to show how Cirq can be used to solve it. We don't include proofs, but we refer the reader to corresponding lemmas in the original paper."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Preparation and bruteforce solution\n",
+    "\n",
+    "For small values of $n$ we can solve this problem with a trivial bruteforce solution. First, we need to build $\\mathcal{L}_q$ by checking for all $2^n$ binary vectors, which of them belongs to it (by definition). Then we need to try all possible $z \\in \\mathbb{F}_2^n$, and for each of them and for each $x \\in \\mathcal{L}_q$ check whether $q(x) = 2 z^T x$.\n",
```

Comment:
"which of them belongs to it" is confusing.  I am not sure what you mean by this phrase.

Reply:
Rephrased.